spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,384 @@
+  counterpartyModuleIdentifier: Identifier
+  nextSequenceSend: uint64
+  nextSequenceRecv: uint64
+}
+```
+
+An IBC *packet* is a particular datagram, defined as follows:
+
+```typescript
+interface Packet {
+  sequence: uint64
+  timeoutHeight: uint64
+  sourceConnection: Identifier
+  sourceChannel: Identifier
+  destConnection: Identifier
+  destChannel: Identifier
+  data: bytes
+}
+```
+
+### Desired Properties
+
+#### Efficiency
+
+- The speed of packet transmission and confirmation should be limited only by the speed of the underlying chains.
+
+#### Exactly-once delivery
+
+- Exactly-once packet delivery, without assumptions of network synchrony and even if one or both of the chains should halt (no-more-than-once delivery in that case).
+
+#### Ordering
+
+- Ordering, for ordered channels, whereby if packet *x* is sent before packet *y* on chain `A`, packet *x* must be received before packet *y* on chain `B`.
+
+IBC channels implement a vector clock for the restricted case of two processes (in our case, blockchains). Given *x* → *y* means *x* is causally before *y*, chains `A` and `B`, and *a* ⇒ *b* means *a* implies *b*:
+
+Every transaction on the same chain already has a well-defined causality relation (order in history). IBC provides an ordering guarantee across two chains which can be used to reason about the combined state of both chains as a whole.
+
+For example, an application may wish to allow a single tokenized asset to be transferred between and held on multiple blockchains while preserving fungibility and conservation of supply. The application can mint asset vouchers on chain `B` when a particular IBC packet is committed to chain `B`, and require outgoing sends of that packet on chain `A` to escrow an equal amount of the asset on chain `A` until the vouchers are later redeemed back to chain `A` with an IBC packet in the reverse direction. This ordering guarantee along with correct application logic can ensure that total supply is preserved across both chains and that any vouchers minted on chain `B` can later be redeemed back to chain `A`.
+
+#### Permissioning
+
+- Channel ends should be permissioned to one module on each end.
+
+## Technical Specification
+
+![channel-state-machine](channel-state-machine.png)
+
+### Channel opening handshake
+
+```typescript
+interface ChanOpenInit {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  counterpartyChannelIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+}
+```
+
+```typescript
+function chanOpenInit() {
+  moduleIdentifier = getCallingModule()
+  assert(get("connections/{connectionIdentifier}/channels/{channelIdentifier}") === null)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  channel = Channel{INIT, moduleIdentifier, counterpartyModuleIdentifier, counterpartyChannelIdentifier, 0, 0}
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
```

Comment:
Isn't channelIdentifier an "identifier"? This code says it is a "Channel" (??)

Reply:
Hmm, where does this code say that?